### ISSUE TITLE
Using GraphRAG for Automated Data Analysis with LLMs to Explore Cancer Pharmacogenomics Data

### ISSUE THREAD

**Issue by cannin** [nrnb/GoogleSummerOfCode#260]
### Background
CellMiner Cross-Database (CellMinerCDB, discover.nci.nih.gov/cellminercdb) allows integration and analysis of molecular and pharmacological data within and across cancer cell line datasets. The software is written in the R programming language using the R Shiny web framework. The database contains data on -omics (e.g., gene expression, mutations, etc.), drug response, and sample annotation information that internally exist as tables that have been processed to simplify data analysis. The web interface a set of analyses to be performed with accompanying plots, but it is limited. The amount of data across individual datasets is also growing. The goal is to create tools that leverage large language models (LLMs) to explore this data from a chat interface. 

### Goal
The goal of the project develop a chat interface using LLM-based AI for exploring the available data. The interface will need to interpret user instructions to generate working code making use of the CellMinerCDB API (rcellminer, see related links) that will then be run with a response returned. The data within the CellminerCDB database is amendable to mapping as a graph for use with GraphRAG (e.g., 1: publications describe datasets, genes, drugs, samples, 2: drugs target proteins, etc.) for data retrieval. 

### Getting Started
Look at the links; read the CellMinerCDB publication and rcellminer documentation. Try to find similar projects or ideas online using LangChain. Draft a proposal with your ideas.

### Difficulty Level: Medium
Simple to start, but a major challenge is in getting "working code" that necessitates efforts such as advanced prompt engineering and other techniques the candidate needs to identify.

### Size and Length of Project
- **medium: 175 hours**
- **12 weeks**

### Skills
* Essential skills: Python
* Nice to have skills: R, LangChain, LangGraph, data analysis, and LLM APIs (GPT4, Ollama, etc.), Neo4J/Cypher

# Related links
* https://www.langchain.com/langchain
* https://github.com/CBIIT/cellminercdb
* https://github.com/CBIIT/rcellminer
* https://www.bioconductor.org/packages/release/bioc/html/rcellminer.html
* https://bioconductor.org/packages/release/data/experiment/html/rcellminerData.html
* https://discover.nci.nih.gov/rsconnect/cellminercdb/
* Publication with example analyses to be automated: https://academic.oup.com/nar/article/49/D1/D1083/5983630

### Potential Mentors
Augustin Luna

**Comment by Neilblaze** [comment]
Hi @cannin, although I understand that NRNB is currently awaiting approval from GSoC admin, this project seems very promising, and I'd love to contribute to this project if it's up for grabs. I'm a grad student in CS located in IST (+05:30). As per the compatibility, I have experience working with most of the mentioned skill sets, including Python (~2 years), and have a good hold on data analysis, LLMs, Prompt Engineering + Designing, RAG, Knowledge Graphs, etc. from my past research and hackathon experiences.

Please let me know the next steps to help me get started. Thanks! 🙏🏼 

---

Also, in the meantime, I'd like to know the motivation behind opting for [GraphRAG](https://github.com/microsoft/graphrag) and not [LightRAG](https://lightrag.github.io/), as LightRAG is comparatively much faster, more cost-efficient (i.e. affordable) and most importantly _allows incremental updates to graphs without full regeneration_, which saves time, every time the graph is recursively updated, which is very essential to ensure high throughput & consistency (of newly inserted data) assuming the chat system will be used by many at any instant of time. Is there any reference design doc, that I can look upto to get a holistic picture of the system?

**Comment by xts-Michi** [comment]
Hey @cannin,

This project looks super interesting, and I'd love to contribute! I’m a Bioinformatics student with experience in Python and currently learning R, which I should be comfortable with by the time GSoC starts. I’ve worked with LLMs privately and have a strong interest in AI and Machine Learning.
Previously, I contributed to TFpredict, where we processed large biological datasets, so I’m comfortable working with structured data. This project excites me because it combines AI with bioinformatics, which is exactly the kind of challenge I enjoy.
I’d love to get started—are there any tips for familiarizing myself with CellMinerCDB, or should I just explore the repo and docs on my own?

Looking forward to your response!

**Comment by cannin** [comment]
@Neilblaze @xts-Michi I have added a Getting Started section.

**Comment by Neilblaze** [comment]
@cannin Perfect, thanks! 🙏🏼

**Comment by xts-Michi** [comment]
@cannin I was just told by Dr. Dräger that we already met at the COMBINE conference in Stuttgart in 2024. Maybe you remember that we talked about Systems Biology and the TFpredict project I presented?
I would also like to ask what the next steps are for the application and how you could become my mentor. I’m very interested and would love to know the concrete next steps so that I can apply for GSoC

**Comment by cannin** [comment]
@xts-Michi thanks. check out Getting Started section.

**Comment by ayush1009208** [comment]
Hello @cannin can i work on this?

**Comment by khanspers** [comment]
NRNB has been accepted as a mentoring organization for GSoC 2025. **The contributor application period is March 24 – April 8**. Here are some useful links:

[GSoC contributor guide](https://google.github.io/gsocguides/student/)
[NRNB project proposal template](https://docs.google.com/document/d/1Zi6L38CHEeq2aL6xzv0Ozhd_Y6D71W3yCBGHplmxr6k/edit?usp=sharing)
[Eligibility requirements](https://developers.google.com/open-source/gsoc/faq#what_are_the_eligibility_requirements_for_participation)
[Full program timeline](https://developers.google.com/open-source/gsoc/timeline)

**Comment by AryanPrakhar** [comment]
Hi @cannin,  

I've been exploring CellMinerCDB, rcellminer, and GraphRAG, and this project aligns well with my experience. I'm a student at IIT (BHU) Varanasi and have worked on LLM benchmarks (ICLR’25) and multi-agent data-driven scientific discovery (ICML’24), along with open-source contributions through Code4GovTech.  

I’ve started drafting my proposal and would appreciate any feedback. Would you be open to reviewing it? Looking forward to your thoughts!

**Comment by cannin** [comment]
@AryanPrakhar I am willing to review an application draft 1-2 times; aluna@broadinstitute.org send if you think it is advanced enough.

**Comment by sriramsowmithri9807** [comment]
Hi @cannin @khanspers 

I’m really excited about the CellMinerCDB chat interface project and would love to contribute! The idea of using LLMs to simplify complex data exploration is fascinating, and I’m eager to help build a solution that makes the database more accessible.

I’ve started exploring the CellMinerCDB publication, rcellminer docs, and LangChain examples to get up to speed. I’m particularly interested in tackling the challenge of translating user queries into actionable code and leveraging GraphRAG for data retrieval.

Let me know how I can help or if there’s a good starting point to dive in! Looking forward to collaborating.

**Comment by stellaxu2077** [comment]
Hi @cannin,

This project looks fascinating, and I’d love to contribute! I’m a master’s student in bioinformatics at the University of Copenhagen, with experience in deep generative models for cancer multi-omics data analysis. My current research focuses on deep learning models for RNA-binding site prediction, and I have hands-on experience with PyTorch, NumPy, Pandas, and high-performance computing.

I’m also familiar with GitHub workflows and have worked with protein language models especially those based on transformers. Recently, I started learning about knowledge graphs and their applications in biomedical research, which makes me particularly excited about the GraphRAG approach in this project.

I’d love to learn more about how I can get involved. Looking forward to your response!

**Comment by cannin** [comment]
@stellaxu2077 Thanks for the interest see "Getting Started"

**Comment by AnkurKumarShukla** [comment]
### Interest in Developing a Chat Interface for CellMinerCDB Using LLM-Based AI

**Name:** Ankur Kumar Shukla  
**Email:** shuklaankur123456789@gmail.com 
**About Me :** [Portfolio](https://ankurshukla.nextbrains.in) 
**GitHub Profile:** [AnkurKumarShukla](https://github.com/AnkurKumarShukla)  

---

**Introduction:**  
hello @cannin I am writing to express my strong interest in contributing to the "CellMinerCDB Chat Interface Using LLM-Based AI" project. With experience in **hybrid retrieval systems**, **multi-LLM integration**, and **state-of-the-art retrieval mechanisms**, I believe I can make meaningful contributions to this project. My background in AI-driven applications aligns closely with the project's goals, and I am eager to enhance my skills through impactful open-source collaboration.  

---

**Relevant Experience and Skills:**  
I have worked on a **Retrieval-Augmented Generation (RAG)** system using **LangChain** and **LangGraph**, which integrated multiple LLM APIs for effective data retrieval. This experience aligns well with developing an interactive chat interface for CellMinerCDB.  

Additionally, my participation in multiple hackathons has enhanced my ability to develop robust solutions rapidly. My experience with **GraphRAG**, **Neo4J/Cypher**, and **Python** will be valuable for building efficient data retrieval solutions.

---

**Contribution:**  
I aim to contribute by:
- Developing a chat interface using **LangChain** and **LLM APIs**.
-  Working on implementing the most effective way to retrieve the relevant documents associated with a query.
- Implementing effective **prompt engineering** to translate user instructions into accurate code.
- Leveraging **GraphRAG** to map complex data relationships for enhanced retrieval.

---

**Conclusion:**  
I am excited about the opportunity to contribute to this impactful project and enhance my skills through open-source collaboration. I look forward to discussing my application further.  

**Sincerely,**  
Ankur Kumar Shukla  
B.Tech in Computer Science & Engineering  
Indian Institute of Information Technology Vadodara

**Comment by aadyasharma13** [comment]
**Interest in Developing a Chat Interface for CellMinerCDB Using LLM-Based AI**

Name: Aadya Sharma
Email: aadyasindia@gmail.com

⸻

**Introduction:**

Hello @cannin,

I am excited to express my strong interest in contributing to the “CellMinerCDB Chat Interface Using LLM-Based AI” project. With expertise in machine learning, Python, and AI-driven applications, I have worked extensively with LangChain, retrieval-augmented generation (RAG), and multi-LLM integration. My experience in data visualization (D3.js), Go, and computational modeling further strengthens my ability to develop effective solutions for this project.

⸻

**Relevant Experience and Skills:**

I have hands-on experience in developing ML models for medical diagnostics, working with structured and unstructured data to build accurate predictive models. My work on LLM-based chatbot development has strengthened my understanding of context-aware retrieval, prompt engineering, and response optimization.

Some of my key technical skills relevant to this project include:
- Machine Learning & AI: Built predictive models for diabetes and breast cancer detection.
-LLM-Based Chatbots: Developed AI chatbots using LangChain and retrieval-augmented generation (RAG).
-Python & AI Libraries: TensorFlow, PyTorch, scikit-learn, LangChain.
- Data Retrieval & Processing: Vector search (FAISS, Pinecone), Neo4J/Cypher for knowledge graphs.
- Software Development: Backend integration for AI applications, Go, D3.js for data visualization.
⸻

**Proposed Contribution:**
	1.	**Developing a Conversational AI for CellMinerCDB**
	•	Implementing LLM-based AI chat interfaces to allow researchers and clinicians to interact seamlessly with the database.
	•	Ensuring multi-turn conversations for improved user experience and response accuracy.
	2.	**Enhancing Biomedical Data Retrieval**
	•	Developing a hybrid retrieval mechanism (BM25 + Vector Search) for precise document extraction.
	•	Implementing context-aware retrieval for answering user queries with minimal hallucination.
	3.	**Optimizing ML-Driven Query Interpretation**
	•	Fine-tuning prompt engineering and instruction tuning to improve LLM responses.
	•	Utilizing structured queries (Cypher, SQL) for retrieving CellMinerCDB data efficiently.
	4.	**Improving Medical Insights with ML Models**
	•	Integrating biomedical ML models for predictive analytics within the chat interface.
	•	Applying knowledge graphs (Neo4J) to enhance connections between biomarkers, drugs, and disease pathways.
	5.	**Ensuring Scalability & Performance Optimization**
	•	Implementing LangChain agents with memory for maintaining conversation history.
	•	Optimizing latency and response generation using efficient model inference techniques.
	6.	**Enhancing Explainability & Transparency**
	•	Developing interpretable AI responses that provide sources and reasoning behind outputs.
	•	Implementing confidence scores and citations for biomedical queries to ensure trust in AI-generated insights.
	7.	**Deploying & Integrating with CellMinerCDB**
	•	Designing a REST API for easy integration into existing CellMinerCDB infrastructure.
	•	Ensuring the chatbot supports multimodal inputs (text, structured queries, etc.).
⸻

**Conclusion:**

I am eager to contribute to this impactful project, applying my expertise in AI, ML, and data-driven applications to improve the CellMinerCDB platform. Looking forward to discussing how I can support this project further!

Best regards,
Aadya Sharma
https://www.linkedin.com/in/aadya13/

**Comment by Tr3nt-Xie** [comment]
Hi @cannin ,
This project looks interesting and I'd like to contribute. And I have started to write my proposal draft and done some previous work.
And, I’m reaching out to clarify the intended scope of the project, as I’ve encountered some ambiguity while working to structure CellMinerCDB’s complex data into a knowledge graph for Neo4J. My prior experience building a knowledge graph and RAG system for traditional medicine data involved relatively straightforward CSV transformations, but CellMinerCDB presents unique challenges due to its multi-layered heterogeneity—particularly the cross-dataset naming discrepancies for shared cell lines and drugs noted in the documentation.

What remains unclear is whether we’re aiming to develop a comprehensive, general-purpose LLM agent capable of handling arbitrary queries across the entirety of CellMinerCDB (requiring exhaustive modeling of all entities, relationships, and metadata), or a more specialized agent focused on specific workflows like drug-gene correlation analysis (which could justify selective data inclusion). This distinction profoundly impacts both the knowledge graph schema and RAG implementation. For instance, a general-purpose agent would need to embed all cross-dataset mappings as first-class nodes/relationships, whereas a task-specific agent might treat mappings as auxiliary retrieval-augmentation tools.

The existing rcellminerUtilsCDB [https://github.com/CBIIT/rcellminerUtilsCDB](url) mapping tables (e.g., cellLineMatchTab, drugSynonymTab) seem invaluable for disambiguation, but I’d appreciate guidance on whether these should be:

Fully materialized in the graph (potentially bloating the schema but enabling richer traversals), or

Maintained externally and invoked during RAG retrieval to filter/expand queries.

Understanding the project’s ultimate use cases—whether exploratory data access for bench scientists or targeted hypothesis testing—will help me optimize the trade-offs between KG complexity, query latency, and LLM prompt efficiency. Any clarification on prioritization (e.g., “start with drug-cell line responses, then expand to multi-omics”) would be immensely helpful for scoping the MVP.

**Comment by cannin** [comment]
@DickBrain-ray 

## Scope
Obviously more and well done is good, but your project proposal should be feasible in the summer. I link a paper showing example analyses that can help limit scope. 

### More Comments 
* I discovered this project (PandasAI https://github.com/sinaptik-ai/pandas-ai/blob/main/assets/demo.gif) after posting this GSOC idea. The idea is similar; a difference in what is being requested is the "graph" elements that could be used for more advanced querying. While a lot of data for this project (within rcellminerData) is in numeric matrices, there is also a lot of data that is not numeric. There are mapping tables and other annotation tables that could go into the GraphRAG for queries likely "give me all the topoisomerase inhibitors" or "only genes involved in metabolism" that might benefit from GraphRAG and the background knowledge in LLMs:
* Not the original idea, but I would be open to a project proposal that tries to do everything in Python (i.e., exports rcellminerData to Python). 

### Getting Annotation (annot) Tables 
```
library(rcellminer)

# DRUG DATA ----
drugAct <- exprs(getAct(rcellminerData::drugData))
drugAnnot <- getFeatureAnnot(rcellminerData::drugData)[["drug"]]

# MOLECULAR DATA ----
expData <- exprs(rcellminerData::molData[["exp"]])
expAnnot <- getFeatureAnnot(rcellminerData::molData)[["exp"]]
```

**Comment by Tr3nt-Xie** [comment]
@cannin 
Thank you for the detailed feedback and the PandasAI reference—it significantly clarified the project scope and technical expectations. After reviewing their demo, I now better understand how to balance usability with biological specificity in our LLM-agent design. Here's my current progress and some questions for alignment:

**Current Progress**

1. **R-to-Python Data Migration:**

I’ve begun converting rcellminerData into Python-friendly formats (CSV) using rpy2, prioritizing drug annotations (moa, FDA status) and gene-pathway mappings.

2. **Graph Data Structuring:**

Actively developing code to transform key datasets into triples with attributes for Neo4j:

Nodes: Drugs (properties: id, name, moa), Genes (symbol, pathway), CellLines (id, tissue)

Edges: TARGETS (drug→gene, evidence: str), RESPONDS_TO (cell→drug, ic50: float)

Goal: Enable queries like MATCH (d:Drug {moa: "Topoisomerase inhibitor"})-[]->(g:Gene) with minimal ambiguity.

3. **Few-Shot Prompt Engineering:**

After seeing PandasAI's demo and your feedback, I realize the importance of non-numeric annotation data for GraphRAG which insipire me another idea. I've specifically designed my few-shot templates to address exactly these types of biological concept queries. While not core to the pipeline, I’m prototyping LLM prompts to handle synonym mapping (e.g., "Herceptin" → "Trastuzumab") and query disambiguation. Example template:

`User: "Find genes linked to platinum drugs"  
System: SEARCH Drug WHERE moa CONTAINS "platinum"; LINK Gene VIA TARGETS  `

**Questions:**

How should I upload my proposal? And, beyond my current work, what else could I prepare for my proposal to strengthen my application?

**Comment by cannin** [comment]
@DickBrain-ray if you want proposal feedback before deadline email me at aluna@broadinstitute.org with draft

**Comment by ttchristian** [comment]
@cannin Hi. I just wanted to let you know that I’ve sent you an email regarding my proposal. I’d really appreciate it if you could take a moment to review it whenever you get a chance. My email account is ztong@wpi.edu. Thanks in advance!
Best,
Zhongyi Tong

**Comment by aadyasharma13** [comment]
@cannin Hi. I just wanted to let you know that I’ve sent you an email regarding my proposal. I’d really appreciate it if you could take a moment to review it and provide relevant feedback. My email account is aadyasindia@gmail.com . Thanks in advance!
Regards,
Aadya Sharma

**Comment by GGCav** [comment]
@cannin Hi, I have sent you an email with my initial proposal. Would you please take a look of it and provide any suggestions? My email address is jeffreyhe406@gmail.com. Thanks so much!
Best, regards.
Jinfeng He

**Comment by Tr3nt-Xie** [comment]
Hi @cannin ,  

Just a friendly reminder that I sent you an email earlier this week regarding my GSoC proposal draft and progress updates. The email address is fengmaox@usc.edu

Please let me know if you need any additional materials or clarifications.

**Comment by khanspers** [comment]
This project has been accepted for GSoC 2025. This issue will be closed since the project is no longer available to other contributors.
